drivers: interrupt_controller: Add SweRV PIC support

olofk opened the pull request:
Add support for the built-in Programmable Interrupt Controller
found in the SweRV EH1 RISC-V CPU

Signed-off-by: Olof Kindgren <olof.kindgren@gmail.com>

The driver is used in the [SweRVolf SoC](https://github.com/chipsalliance/Cores-SweRVolf)

olofk committed:
drivers: interrupt_controller: Add SweRV PIC support

Add support for the built-in Programmable Interrupt Controller
found in the SweRV EH1 RISC-V CPU

Signed-off-by: Olof Kindgren <olof.kindgren@gmail.com>

olofk commented:
Any takers on this? I understand it's a bit niche, but it has battle tested with a bunch of interrupt sources

kgugala reviewed (changes requested):
@olofk, thanks for the contribution. I have some questions and remarks regarding the code.

olofk commented on drivers/interrupt_controller/swerv_pic.c:
```diff
@@ -0,0 +1,282 @@
+	__asm__ swerv_pic_writecsr(meicidpl, 0);
+	__asm__ swerv_pic_writecsr(meicurpl, 0);
+
+	swerv_pic_set_vector(swerv_pic_dummy);
```
This was intended as initial support for the optional hardware vectored interrupts, but this feature is not yet used or tested, so it's better to remove it completely

olofk commented on drivers/interrupt_controller/swerv_pic.c:
```diff
@@ -0,0 +1,282 @@
+
+static void swerv_pic_init_core(struct device *dev)
+{
+	ARG_UNUSED(dev);
```
With the set_vector function removed I folded the logic into `swerv_pic_init` and removed this function

olofk commented on drivers/interrupt_controller/swerv_pic.c:
```diff
@@ -0,0 +1,282 @@
+
+	swerv_pic_init_core(dev);
+
+	/* Setup IRQ handler for PLIC driver */
```
Yes. Fixed now

olofk commented on drivers/interrupt_controller/swerv_pic.c:
```diff
@@ -0,0 +1,282 @@
+		    NULL,
+		    0);
+
+	/* Enable IRQ for PLIC driver */
```
Yes. Fixed now

olofk commented:
Many thanks for the review @kgugala . I have fixed the issues you reported, rebased against master and pushed a new version. The following changes are also made to the new version

* Drop the logging code. It seems no other interrupt drivers have logging and I guess it's a bit awkward to log inside an interrupt routine anyway
* Converted all `z_arch` to `arch` to follow the API namespace changes

kgugala commented on drivers/interrupt_controller/swerv_pic.c:
```diff
@@ -0,0 +1,235 @@
+{
+	u32_t key;
+
+	if (irq <= RISCV_MAX_GENERIC_IRQ)
```
please add braces here https://docs.zephyrproject.org/latest/contribute/index.html#coding-style

olofk replied:
Thanks. Looks like both checkpatch and I missed that one. Fixed, rebased and pushed

kgugala reviewed (changes requested):
Small formatting comment, otherwise LGTM

olofk commented:
@npitre @ioannisg @andrewboie @galak @nategraff-sifive  Any updates on this?

npitre commented:
On Wed, 4 Dec 2019, Olof Kindgren wrote:

> @npitre @ioannisg @andrewboie @galak @nategraff-sifive  Any updates on this?

I do not have experience with Zephyr IRQ driver so I don't have much to contribute here.
